p2p/http/libp2phttp.go

```diff
@@ -434,16 +436,50 @@ func (rt *streamRoundTripper) RoundTrip(r *http.Request) (*http.Response, error)
 		}
 	}()
 
-	// TODO: Adhere to the request.Context
-	resp, err := http.ReadResponse(bufio.NewReader(s), r)
+	var rdr io.Reader
+	if r.Context() == context.Background() {
+		rdr = s
+	} else {
+		// Adhere to the request.Context if context is cancelable.
+		rdr = &readerCtx{
+			ctx: r.Context(),
+			s:   s,
+		}
+	}
+
+	resp, err := http.ReadResponse(bufio.NewReader(rdr), r)
 	if err != nil {
+		s.Close()
 		return nil, err
 	}
 	resp.Body = &streamReadCloser{resp.Body, s}
 
 	return resp, nil
 }
 
+type readerCtx struct {
+	ctx context.Context
+	s   network.Stream
+}
+
+func (r *readerCtx) Read(p []byte) (int, error) {
+	var n int
+	var err error
+	done := make(chan struct{})
+	go func() {
```

Comment:
I'm not sure I want to spawn a go routine for every read call. It might be better to use the result of `ctx.Deadline()` to set the read deadline